Kernel/Process.cpp

```diff
@@ -1551,6 +1556,127 @@ void Process::crash(int signal, u32 eip)
+    // by creating a PT_NOTE Phdr for every threa and filling
+    // in the appropriate fields.
+
+    // Dump all regions to file
+    for (Region& region : m_regions) {
+        Elf32_Phdr phdr;
+        VMObject& vmobj = region.vmobject();
+
+        phdr.p_type = PT_LOAD;
+        phdr.p_offset = offset;
+        phdr.p_vaddr = reinterpret_cast<uint32_t>(region.vaddr().as_ptr());
+        phdr.p_paddr = 0;
+
+        // This is a bit janky because we need to work out how big the memory
+        // region will be in the file. This sadly can't just be region.size()
+        // because it will mess up the size of the file. GDB can detect, and
+        // WILL complain about this.
+        size_t mapped_page_count = 0;
+        for (size_t page_num = 0; page_num < vmobj.page_count(); page_num++) {
+            PhysicalPage* page = vmobj.physical_pages()[page_num];
+            if (!page)
+                continue;
+
+            mapped_page_count++; // Only consider mapped pages (so we don't go past the size of `region`)
+        }
+
+        // FIXME: memsz should be `region.page_count() * PAGE_SIZE` according
+        // to the ELF spec, but that causes readelf to issue a warning about filesz
+        // being greater than memsz!!??? How is that even possible??
+        phdr.p_filesz = mapped_page_count * PAGE_SIZE;
+        phdr.p_memsz = mapped_page_count * PAGE_SIZE;
+        phdr.p_align = 0;
+
+        phdr.p_flags = region.is_readable() ? PF_R : 0;
+        if (region.is_writable())
+            phdr.p_flags |= PF_W;
+        if (region.is_executable())
+            phdr.p_flags |= PF_X;
+
+        offset += phdr.p_filesz;
+
+        fd->write(reinterpret_cast<uint8_t*>(&phdr), sizeof(Elf32_Phdr));
+    }
+
+    // Get all the threads so we can write stuff to disk (we can't in this lambda because
+    // interrupts are disabled)
+    Vector<Thread*> threads;
+    current->for_each_thread([&](Thread& thread) {
+        threads.append(&thread);
+        return IterationDecision::Continue;
+    });
+
+    for (size_t i = 0; i < threads.size(); i++) {
+    }
+
+    // Let's actually dump all the memory into the core file
+    for (Region& region : m_regions) {
+        // Remap the region so that it's readable by us.
+        region.set_readable(true);
+        region.remap();
+
+        // Now let's walk the physical pages and see which ones are actually mapped.
+        VMObject& vmobj = region.vmobject();
+        size_t mapped_page_count = 0;
+        for (size_t page_num = 0; page_num < vmobj.page_count(); page_num++) {
+            PhysicalPage* page = vmobj.physical_pages()[page_num];
+            if (!page)
+                continue;
+
+            fd->write(reinterpret_cast<uint8_t*>((region.vaddr().as_ptr() + (mapped_page_count * PAGE_SIZE))), PAGE_SIZE);
+            mapped_page_count++; // Only consider mapped pages (so we don't go past the size of `region`)
+        }
+    }
+    fd->close();
```

Comment:
This doesn't do anything. You can leave it out and `~FileDescription()` will take care of everything.

Reply:
It kind of does, it's supposed to flush enough caches to check for errors. It doesn't do it today, but checking for success at call site (so here) wouldn't hurt. If the call site was to check, that is.

Reply:
This call site does not communicate anything to userspace anyway, so all we could do is log some whiny message?

Reply:
Yes.